junit-jupiter-api/src/main/java/org/junit/jupiter/api/DynamicNode.java

```diff
@@ -29,8 +32,12 @@
 
 	private final String displayName;
 
-	DynamicNode(String displayName) {
+	/** Custom test source instance associated with this node; potentially {@code null}. */
+	private final TestSource testSource;
```

Comment:
Alarms going off....

Unless I'm missing something, the direct dependency on `TestSource` in a field makes **all** dynamic containers and dynamic tests dependent on the `junit-platform-engine` module.

Consequently, the "optional" dependency on `junit-platform-engine` is meaningless.

In other words, with the hard dependency on `TestSource` it will be impossible to use dynamic tests without having the `junit-platform-engine` in the classpath.

So, @junit-team/junit-lambda, we need to rethink this.

Reply:
This brings us back to my original comment on the matter: https://github.com/junit-team/junit5/issues/1178#issuecomment-346371161

> providing support for something like `TestSource` would require a new API that can be converted into `TestSource`.

Reply:
Never mind: as @sormuras pointed out, the `TestSource` is indeed a run-time dependency but not necessarily a compile-time dependency. So that should allow it to be "optional".

But let's add some integration tests to make sure all is well. 😉